### PR TITLE
fix: Table editor deleting / pagination bug

### LINKED ISSUE #9305: Table editor deleting / pagination bug
# Bug report

Deleting 100 rows in the table leaves the table in a weird state, requiring a refresh.  At first, it looks like you've deleted all of your rows, not just the 100 you selected. 

**To reproduce**

Create a new table with a lot of rows:

1. Go to the SQL editor → Welcome page → Quick start section → Countries to quickly create a table with a lot of entries
2. Go to that new countries table
3. Select all rows at the top, then Delete 100 rows
 
CleanShot 2022-10-03 at 13 07 32@2x
4. Notice that your table empty. It looks like the entire table content has been deleted, but that's not the case. 
CleanShot 2022-10-03 at 13 18 02@2x

### PR DESCRIPTION
**Pull request by gitstart**
## What kind of change does this PR introduce?

Fixes #9305

## Bug Fix
Table should reload with the correct row data once all the records on the current page is deleted.
## What is the current behavior?
Deleting 100 rows in the table leaves the table in a weird state, requiring a refresh. At first, it looks like you've deleted all of your rows, not just the 100 you selected.

## What is the new behavior?
Table loads with the correct data rows once the first 100 rows are loaded without requiring a refresh.

## Demo
https://user-images.githubusercontent.com/51731962/233307574-22e3542b-29e9-4924-8afe-da12e73a1830.mov

---
This code was written and reviewed by GitStart Community. Growing future engineers, one PR at a time.

### CODE REVIEW

**Comment by gitstart** [comment]
Hi @saltcod kindly help take a look at solution added on this PR.

File: studio/components/grid/store/reducers/row.ts [hunk 1]
```diff
@@ -142,6 +142,7 @@ const RowReducer = (state: RowInitialState, action: ROW_ACTIONTYPE) => {
         ...state,
         rows: state.rows.filter((x) => !action.payload.rowIdxs.includes(x.idx)),
         totalRows: totalRows,
+        refreshPageFlag: Number(state.totalRows > state.page),
```
Comment (joshenlim) [inline studio/components/grid/store/reducers/row.ts:145]: can i check in on this - what's the reason behind `Number(state.totalRows > state.page)`?

this fix does work, but i dont quite understand why. Shouldn't we just assign `refreshPageFlag: REFRESH_PAGE_IMMEDIATELY`?

Comment (gitstart) [reply studio/components/grid/store/reducers/row.ts]: Thanks for the suggestion @joshenlim.
Pushed an update now, to improve readability as you have suggested. Please take a look.

**Review by joshenlim — APPROVED** [review]

**Comment by joshenlim** [comment]
hey there! so sorry i thought i had clicked submit previously when writing the comment

one more issue i found - if your table has more than 1 page, and you go to the last page and delete all the rows, the page ends up showing 0 rows, ideally it would go back to the previous page

**Comment by joshenlim** [comment]
@gitstart hey there! again, so sorry for the delay on this, would you mind pulling the latest changes from master and resolving the conflicts? i'll take a look again once thats done!

**Comment by joshenlim** [comment]
i just found another issue sorry
this query is being run when i delete all the rows on the first page of a table
```select * from public.countries limit 100 offset -100;```

but the offset value should not be negative

File: studio/components/grid/store/reducers/row.ts [hunk 2]
```diff
@@ -136,14 +136,6 @@ const RowReducer = (state: RowInitialState, action: ROW_ACTIONTYPE) => {
...
       }
     }
-    case 'REMOVE_ROWS': {
-      const totalRows = state.totalRows - action.payload.rowIdxs.length
-      return {
-        ...state,
-        rows: state.rows.filter((x) => !action.payload.rowIdxs.includes(x.idx)),
-        totalRows: totalRows,
-      }
-    }
```
Comment (gitstart) [inline studio/components/grid/store/reducers/row.ts:146]: @joshenlim when `REMOVE_ROWS` is dispatched, `rows` and `totalRows` are already updated.

**Comment by alaister** [comment]
Hey @gitstart,

Sorry for all the delays on this one.

It looks like this is no longer a bug in production, so I'm going to close this. Please feel free to reopen if you can still reproduce it.

Thanks for your work 🙏🏻

**Comment by gitstart** [comment]
Hi @alaister 
Thanks for taking a look at this. After testing the main/prod branch again, we noticed that most of the previous cases of reproducing this bug is actually fixed as you have mentioned.
However we also noticed a particular case that can be reproduced when we delete all items on the `last` pagination page, then it doesn't update as expected

Here's a quick demo video to highlight this behavior
https://user-images.githubusercontent.com/51731962/233307574-22e3542b-29e9-4924-8afe-da12e73a1830.mov

Here's a demo of the same behavior on this branch
https://user-images.githubusercontent.com/51731962/233308442-f9a13a93-bfc8-403a-b0dc-2bc92f58e5dd.mov

I currently don't have the permissions to re-open this PR, but would suggest to do so. WDYT?

**Review by alaister — APPROVED** [review]
Hey @gitstart,

Thanks for taking the time to make sure this issue was fully solved. Looks great!

**Pull request merged**